src/MagentoHackathon/Composer/Magento/Repository/InstalledPackageFileSystemRepository.php

```diff
@@ -145,20 +121,15 @@ public function add(InstalledPackage $package)
     {
         $this->load();
 
-        $this->hasChanges = true;
-
         try {
             $this->findByPackageName($package->getName());
-            foreach ($this->packages as &$installedPackage) {
-                if ($installedPackage->getName() === $package->getName()) {
-                    $installedPackage = $package;
-                }
-            }
         } catch (\Exception $e) {
             $this->packages[] = $package;
+            $this->hasChanges = true;
             return;
         }
 
+        throw new \Exception(sprintf('Package: "%s" is already installed', $package->getName()));
```

Comment:
Yes, but when an existing package is updated, it is first removed from the repository, then the new version added. The exception is not thrown. The add function should never be called for the same package more than once.

`findByPackageName` throws an exception if the package doesn't exist. We catch it, add the packages and mark `$hasChanges` to true.